source/opt/unify_const.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "passes.h"
+
+#include <unordered_set>
+
+#include "def_use_manager.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// Using a trie node to store encountered constants and look up later with a
+// given constant defining instruction's opcode, result type id and operands.
+// Each ResultIdTrieNode stores a result id, if it does represent an
+// encountered constant. ResultIdTrieNode also guide the lookup process by
+// holds the map toward the next ResultIdTrieNode to check with.
+struct ResultIdTrieNode {
+  ResultIdTrieNode() : result_id(0), next() {}
+  uint32_t result_id;
+  std::unordered_map<uint64_t, std::unique_ptr<ResultIdTrieNode>> next;
+};
+
+// A helper function to extract the operand's raw value in 64-bit unsigned
+// integers. Note that this does not support operands longer than 64-bit.
+uint64_t GetOperandValueInUint64(const ir::Operand& operand) {
```

Comment:
Updated in 5d7fa1e.

Actually we should not need to care the length of an operand at all....